src/main/java/org/broadinstitute/hellbender/tools/walkers/rnaseq/GeneExpressionEvaluation.java

```diff
@@ -0,0 +1,667 @@
+            } else {
+                if (!read.hasAttribute(SAMTag.MC.toString())) {
+                    throw new GATKException("Mate cigar must be present.  Consider running FixMateInformation.");
+                }
+                final Cigar mateCigar = TextCigarCodec.decode(read.getAttributeAsString(SAMTag.MC.toString()));
+                ret = ret && read.getStart() <= read.getMateStart() + mateCigar.getReferenceLength();
+            }
+        }
+        return ret;
+    }
+
+    static List<Interval> getAlignmentIntervals(final GATKRead read, final boolean spliced, final int minimumMappingQuality) {
+
+        if (spliced) {
+            final List<Interval> alignmentIntervals = new ArrayList<>();
+
+            final List<AlignmentBlock> readAlignmentBlocks = SAMUtils.getAlignmentBlocks(read.getCigar(), read.getStart(), "read cigar");
+
+            for( final AlignmentBlock block : readAlignmentBlocks) {
+                alignmentIntervals.add(new Interval(read.getContig(), block.getReferenceStart(), block.getReferenceStart()+block.getLength() - 1));
+            }
+
+            boolean overlapsMate = false;
+            if (inGoodPair(read, minimumMappingQuality)) {
+                final String mateCigarString = read.getAttributeAsString(SAMTag.MC.toString());
+                if(mateCigarString == null) {
+                    throw new GATKException("Mate cigar must be present if using spliced reads");
+                }
+                final List<AlignmentBlock> mateAlignmentBlocks = SAMUtils.getAlignmentBlocks(TextCigarCodec.decode(mateCigarString), read.getMateStart(), "mate cigar");
+                for( final AlignmentBlock block : mateAlignmentBlocks) {
+                    final Interval alignmentBlockInterval = new Interval(read.getMateContig(), block.getReferenceStart(), block.getReferenceStart()+block.getLength() - 1);
+                    alignmentIntervals.add(alignmentBlockInterval);
+
+                    if (!overlapsMate && read.overlaps(alignmentBlockInterval)) {
+                        overlapsMate = true;
+                    }
+                }
+            }
+            return getMergedIntervals(alignmentIntervals);
+
+        } else {
+            if (read.isUnmapped()) {
+                return Collections.emptyList();
+            }
+
+            final boolean inGoodPair = inGoodPair(read, minimumMappingQuality);
+
+            final int start = inGoodPair? Math.min(read.getStart(), read.getMateStart()) : read.getStart();
+            final int end = inGoodPair? start + Math.abs(read.getFragmentLength()) - 1 : read.getEnd();
+            return Collections.singletonList(new Interval(read.getContig(), start, end));
+        }
+
+
+    }
+
+    @Override
+    public void apply(GATKRead read, ReferenceContext referenceContext, FeatureContext featureContext) {
+        if ((!read.isReverseStrand() || !inGoodPair(read, mappingQualityFilter.minMappingQualityScore))) {
```

Comment:
The idea is that for good pairs, we only need to look at one of the reads, since we can get its mate information from it.  However, for non good pairs, we want to consider each read separately.  For example if we are using unspliced data and the two reads aligned 10000 bases away from each other, and is thus marked as not properly paired, we don't want to consider that a fragment that is 10000 bases long and thus covers every gene between the two alignments.
I suppose it would also be reasonable (as I think is your point), to just ignore all data that isn't properly paired on the assumption that something funny has happened with those reads, and so they shouldn't be trusted.  So I will add an option to use "non good pair" data or not.

Reply:
I don't think you necessarily need to add that option. It's hard to say without looking at data if you should be ignoring "non good" pairs completely or not, so I'd wait until someone asks for that specific use case (unless you've already done it because I've commented here too late 😃).

Reply:
Yeah, I was also just realizing as I was thinking about how to implement it that the effect can mostly be achieved through standard command line read-filters, so I'm going to leave as-is for now.